src/Kinto.elm

```diff
@@ -0,0 +1,316 @@
+    { errno : Int
+    , message : String
+    , code : Int
+    , error : String
+    }
+
+
+type alias StatusCode =
+    Int
+
+
+type alias StatusMsg =
+    String
+
+
+type Error
+    = ConfigError String
+    | ServerError StatusCode StatusMsg String
+    | KintoError StatusCode StatusMsg ErrorRecord
+    | NetworkError Http.RawError
+
+
+
+-- Making requests
+
+
+endpointUrl : Config -> Endpoint -> Url
+endpointUrl config endpoint =
+    let
+        baseUrl =
+            if String.endsWith "/" config.baseUrl then
+                String.dropRight 1 config.baseUrl
+            else
+                config.baseUrl
+
+        joinUrl =
+            String.join "/"
+    in
+        case endpoint of
+            RootEndpoint ->
+                -- Otherwise Kinto returns a 307
+                -- See https://github.com/Kinto/kinto/issues/852
+                baseUrl ++ "/"
+
+            BucketEndpoint Nothing ->
+                joinUrl [ baseUrl, "buckets" ]
+
+            BucketEndpoint (Just bucketName) ->
+                joinUrl [ baseUrl, "buckets", bucketName ]
+
+            CollectionEndpoint bucketName Nothing ->
+                joinUrl [ baseUrl, "buckets", bucketName, "collections" ]
+
+            CollectionEndpoint bucketName (Just collectionName) ->
+                joinUrl [ baseUrl, "buckets", bucketName, "collections", collectionName ]
+
+            RecordEndpoint bucketName collectionName Nothing ->
+                joinUrl [ baseUrl, "buckets", bucketName, "collections", collectionName, "records" ]
+
+            RecordEndpoint bucketName collectionName (Just recordId) ->
+                joinUrl [ baseUrl, "buckets", bucketName, "collections", collectionName, "records", recordId ]
+
+
+client : Config -> Endpoint -> Verb -> Task Error Json.Value
+client config endpoint verb =
+    let
+        request =
+            { verb = verb
+            , url = endpointUrl config endpoint
+            , headers = config.headers
+            , body = Http.empty
+            }
+    in
+        (Http.send Http.defaultSettings request)
+            |> toKintoResponse
+
+
+withHeader : String -> String -> Config -> Config
+withHeader name value ({ headers } as config) =
+    let
+        newHeaders =
+            ( name, value ) :: headers
+    in
+        { config | headers = newHeaders }
+
+
+withAuthHeader : Auth -> Config -> Result Error Config
+withAuthHeader auth config =
+    case auth of
+        NoAuth ->
+            Ok config
+
+        Basic username password ->
+            let
+                hash str =
+                    Base64.encode str |> Result.formatError ConfigError
+
+                addAuthHeader hash =
+                    Ok (withHeader "Authorization" ("Basic " ++ hash) config)
+            in
+                hash (username ++ ":" ++ password)
+                    `Result.andThen` addAuthHeader
+
+        Bearer token ->
+            Ok (config |> withHeader "Authorization" ("Bearer " ++ token))
```

Comment:
Should this be smart enough to remove existing `Authorization` headers? Actually, thinking about it more, should `Auth` be a field in `Config`, rather than the lower-level `headers`?